### PR TITLE
mailbox: add EXTEND_PCR command

### PR DESCRIPTION
**Pull request by leongross**
Initial implementation for the [`EXTEND_PCR`](https://github.com/chipsalliance/caliptra-sw/tree/main/runtime#extend_pcr) mailbox command.

### CODE REVIEW

File: common/src/mailbox_api.rs [hunk 1]
```diff
@@ -276,6 +277,29 @@ impl Default for InvokeDpeReq {
     }
 }
 
+// EXTEND_PCR
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes, PartialEq, Eq)]
+pub struct ExtendPcrReq {
+    pub hdr: MailboxReqHeader,
+    pub pcr_idx: u32,
+    pub value: [u8; ExtendPcrReq::DATA_MAX_SIZE], // variable length
```
Comment (bluegate010) [inline common/src/mailbox_api.rs:286]: It doesn't seem like this is variable-length? `ExtendPcrReq::read_from()` will fail if the caller doesn't pass in 512 bytes of data.

**Review by bluegate010 — COMMENTED** [review]
Can we add test coverage for this?

File: common/src/mailbox_api.rs (hunk 1 above)
Comment (jhand2) [reply common/src/mailbox_api.rs:286]: I think we may need a size field here to tell RT how much of `value` to interpret. Also same for StashMeasurementReq

**Comment by leongross** [comment]
Also another topic to discuss: in the implementation of the [sha384](https://github.com/chipsalliance/caliptra-sw/blob/main/drivers/src/sha384.rs#L142) there is a check that I quite don't understand:


```rust
pub fn pcr_extend(&mut self, id: PcrId, data: &[u8]) -> CaliptraResult<()> {
    let total_bytes = data.len() + SHA384_HASH_SIZE;
        if total_bytes > (SHA384_BLOCK_BYTE_SIZE - 1) {
            return Err(CaliptraError::DRIVER_SHA384_MAX_DATA_ERR);
        }
```

Why is it problematic, that `data.len() + 48 > 127`? Shouldn't this code check if it's `> 128` which is the maximum SHA384 block size `SHA384_BLOCK_BYTE_SIZE`?
This check implies, that the max size for `data` is `128 - 1 - 48 = 79`, is this intended, or am I maybe misunderstanding something?

**Comment by bluegate010** [comment]
It's not clear from reading the [HW specification](https://github.com/chipsalliance/caliptra-rtl/blob/main/docs/Caliptra_Hardware_Specification.pdf) whether multiple blocks would be supported, worth trying.

File: runtime/src/pcr.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    pub(crate) fn execute(drivers: &mut Drivers, cmd_args: &[u8]) -> CaliptraResult<MailboxResp> {
+        let cmd =
+            ExtendPcrReq::read_from(cmd_args).ok_or(CaliptraError::RUNTIME_INSUFFICIENT_MEMORY)?;
+
+        let idx =
+            u8::try_from(cmd.pcr_idx).map_err(|_| CaliptraError::RUNTIME_MAILBOX_INVALID_PARAMS)?;
+        let pcr_index: PcrId =
+            PcrId::try_from(idx).map_err(|_| CaliptraError::RUNTIME_MAILBOX_INVALID_PARAMS)?;
+
+        drivers.pcr_bank.extend_pcr(
```
Comment (bluegate010) [inline runtime/src/pcr.rs:19]: Will also want to append an entry to the PCR log.

File: runtime/src/pcr.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+pub struct ExtendPcrCmd;
+impl ExtendPcrCmd {
+    pub(crate) fn execute(drivers: &mut Drivers, cmd_args: &[u8]) -> CaliptraResult<MailboxResp> {
+        let cmd =
+            ExtendPcrReq::read_from(cmd_args).ok_or(CaliptraError::RUNTIME_INSUFFICIENT_MEMORY)?;
+
+        let idx =
+            u8::try_from(cmd.pcr_idx).map_err(|_| CaliptraError::RUNTIME_MAILBOX_INVALID_PARAMS)?;
+        let pcr_index: PcrId =
+            PcrId::try_from(idx).map_err(|_| CaliptraError::RUNTIME_MAILBOX_INVALID_PARAMS)?;
```
Comment (bluegate010) [inline runtime/src/pcr.rs:17]: May want a more descriptive error code. And tests for the error cases.

File: runtime/src/pcr.rs (hunk 2 above)
Comment (bluegate010) [inline runtime/src/pcr.rs:19]: There's not a _strong_ reason, but it seems like a good idea to reserve PCRs 0-3 for Caliptra, and not allow external callers to extend those. But I'm open to leaving it alone.

File: runtime/tests/integration_tests.rs [hunk 4]
```diff
@@ -827,3 +827,31 @@ fn test_idev_id_cert() {
...
+    );
+
+    let cmd = ExtendPcrReq::new(
+        MailboxReqHeader { chksum: checksum },
+        cmd.pcr_idx,
+        cmd.data_size,
+        cmd.data,
+    );
+
+    let res = model.mailbox_execute(u32::from(CommandId::EXTEND_PCR), &cmd.as_bytes());
```
Comment (bluegate010) [inline runtime/tests/integration_tests.rs:855]: Can we also test other edge cases, like len=0 or 255, bad PCR index, etc.?

File: common/src/mailbox_api.rs [hunk 5]
```diff
@@ -276,6 +277,36 @@ impl Default for InvokeDpeReq {
...
 
+// EXTEND_PCR
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes, PartialEq, Eq)]
+pub struct ExtendPcrReq {
+    pub hdr: MailboxReqHeader,
+    pub pcr_idx: u32,
+    pub data_size: u32,
+    pub data: [u8; ExtendPcrReq::DATA_MAX_SIZE],
+    _pad: [u8; 4 - (ExtendPcrReq::DATA_MAX_SIZE % 4)],
```
Comment (bluegate010) [inline common/src/mailbox_api.rs:288]: If DATA_MAX_SIZE % 4 == 0, _pad should be zero length. Wrap in an outer %?

File: common/src/mailbox_api.rs [hunk 6]
```diff
@@ -276,6 +277,36 @@ impl Default for InvokeDpeReq {
...
+}
+// No command-specific output args
+impl ExtendPcrReq {
+    pub const DATA_MAX_SIZE: usize = sha384::SHA384_BLOCK_BYTE_SIZE - sha384::SHA384_HASH_SIZE - 1;
```
Comment (bluegate010) [inline common/src/mailbox_api.rs:292]: It seems like this leaks HW details that would be nicer to hide. Is there a need to allow the data size to exceed 48 bytes? Alternatively, could we try rejiggering the PCR driver to handle multiple blocks? It's not clear to me that the current limitation is HW-based.

File: runtime/tests/integration_tests.rs (hunk 4 above)
Comment (bluegate010) [reply runtime/tests/integration_tests.rs:855]: Should also test that the data we meant to extend got extended.

File: common/src/mailbox_api.rs (hunk 6 above)
Comment (leongross) [reply common/src/mailbox_api.rs:292]: I think there is a need for the `data_size` attribute to exceed 48 bytes, since `data` is used as an input for the method `PcrBank::extend_pcr`, which can process more than 48 bytes at once in the internal hash digest. To me it looks like this function can take up to `128 - 1 - 48 = 79` bytes as payload (as we already discussed earlier) or am I confusing things here?

Comment (leongross) [reply common/src/mailbox_api.rs:292]: So you propose, that we make the `data_size` and `data` in `ExtendPcrReq` entirely variable (wihtout nay `DATA_MAX_SIZE`) and then handle multiple blocks in the `PcrBank::extend_pcr` method?

Comment (bluegate010) [reply common/src/mailbox_api.rs:292]: More or less. I think we still need some maximum size, but 79 bytes seems fairly arbitrary. Either 48 or 96 or 192 bytes seems a bit cleaner from an API perspective.

Comment (leongross) [reply common/src/mailbox_api.rs]: I see you point, I adjusted the command in such a way, that it should now be possible to pass a variable amount of data, currently limited to 192 bytes as you proposed.

Comment (leongross) [reply common/src/mailbox_api.rs]: Any thoughts on this? @bluegate010

File: common/src/mailbox_api.rs [hunk 7]
```diff
@@ -293,6 +294,45 @@ impl Default for InvokeDpeReq {
...
+pub struct ExtendPcrReq {
+    pub hdr: MailboxReqHeader,
+    pub pcr_idx: u32,
+    pub data_size: u32,
+    pub data: [u8; ExtendPcrReq::DATA_MAX_SIZE],
+    _pad: [u8; 4 - (ExtendPcrReq::DATA_MAX_SIZE % 4) % 4],
+}
+pub type ExtendPcrReqErr = ();
+impl ExtendPcrReq {
+    // pub const DATA_MAX_SIZE: usize = sha384::SHA384_BLOCK_BYTE_SIZE - sha384::SHA384_HASH_SIZE - 1;
```
Comment (bluegate010) [inline common/src/mailbox_api.rs:309]: Let's nix commented-out code.

File: common/src/mailbox_api.rs (hunk 6 above)
Comment (bluegate010) [reply common/src/mailbox_api.rs]: Sorry for the delay! Will resolve this and add further comments inline.

File: common/src/mailbox_api.rs [hunk 8]
```diff
@@ -293,6 +294,45 @@ impl Default for InvokeDpeReq {
...
 
+// EXTEND_PCR
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes, PartialEq, Eq)]
+pub struct ExtendPcrReq {
+    pub hdr: MailboxReqHeader,
+    pub pcr_idx: u32,
+    pub data_size: u32,
+    pub data: [u8; ExtendPcrReq::DATA_MAX_SIZE],
+    _pad: [u8; 4 - (ExtendPcrReq::DATA_MAX_SIZE % 4) % 4],
```
Comment (ArthurHeymans) [inline common/src/mailbox_api.rs:305]: I think extending PCR is typically only done with a digest which has the size of the PCR reg itself?

File: drivers/src/pcr_bank.rs [hunk 9]
```diff
@@ -241,8 +242,43 @@ impl PcrBank {
...
+    ) -> CaliptraResult<()> {
+        let data_block_size = (SHA384_BLOCK_BYTE_SIZE - SHA384_HASH_SIZE - 1) as u32;
+        let blocks: u32 = data.len() as u32 / data_block_size as u32;
+
+        for i in 0..blocks {
+            let d = data
+                .get((i * data_block_size) as usize..((i + 1) * data_block_size) as usize)
+                .ok_or(CaliptraError::DRIVER_PCR_BANK_EXTEND_INVALID_SIZE)?;
+
+            sha.pcr_extend(id, d)?;
```
[the remark this answers is not in the document]
Comment (leongross) [reply drivers/src/pcr_bank.rs:270]: Okay, let's do that.

File: runtime/src/pcr.rs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+        drivers.pcr_bank.extend_pcr_variable(
+            pcr_index,
+            &mut drivers.sha384,
+            &cmd.data
+                .get(..cmd.data_size as usize)
+                .ok_or(CaliptraError::DRIVER_PCR_BANK_EXTEND_INVALID_SIZE)?,
+        )?;
+
+        // 2. Add log entry
```
[the remark this answers is not in the document]
Comment (leongross) [reply runtime/src/pcr.rs:40]: So no logging at all or should we find another location to log this to?

[the remark this answers is not in the document]
Comment (bluegate010) [reply runtime/src/pcr.rs:40]: I'm thinking no logging at all.

File: common/src/mailbox_api.rs [hunk 11]
```diff
@@ -293,6 +294,44 @@ impl Default for InvokeDpeReq {
...
```
Comment (ArthurHeymans) [inline common/src/mailbox_api.rs:305]: only pub data: [u8; 48] is needed. No need for a size nor padding.

File: runtime/tests/integration_tests.rs [hunk 12]
```diff
@@ -895,3 +895,82 @@ fn test_idev_id_cert() {
...
+            cmd.data,
+        )
+    }
+
+    let mut model = run_rt_test(None, None);
+    let payload_data = [0u8; ExtendPcrReq::DATA_MAX_SIZE];
+
+    let cmd = generate_mailbox_extend_pcr_req(4, payload_data.len(), payload_data).unwrap();
+    let res = model.mailbox_execute(u32::from(CommandId::EXTEND_PCR), &cmd.as_bytes());
+    assert!(res.is_ok());
```
[the remark this answers is not in the document]
Comment (leongross) [reply runtime/tests/integration_tests.rs:935]: As I see it this could only be done using the `ureg::Mmio`, which atm is not implemented in the emulator yet, right?

[the remark this answers is not in the document]
Comment (leongross) [reply runtime/tests/integration_tests.rs:935]: @bluegate010 any thoughts on this?

File: common/src/mailbox_api.rs (hunk 11 above)
Comment (ArthurHeymans) [reply common/src/mailbox_api.rs]: Drop data_size too.

File: runtime/tests/integration_tests.rs [hunk 13]
```diff
@@ -895,3 +898,108 @@ fn test_idev_id_cert() {
...
+        182, 215, 239, 9, 244, 123, 192, 213, 72, 247, 223, 97, 50, 203, 98, 85, 255, 185, 136,
+        117, 132, 45, 47, 144, 171, 30, 11, 237, 104, 239, 127, 74, 45, 22, 202, 113, 89, 231, 46,
+        208, 185, 237, 115, 123, 223, 231, 246, 254, 103, 112, 39, 33, 159, 23, 201, 17, 64, 144,
+        24, 179, 34, 251, 43, 5, 215, 169, 71, 118, 156, 103, 123, 101, 170, 5, 38, 190, 72, 75,
+        81, 223, 72, 131, 215, 100, 50, 150, 57, 34, 255, 172, 39, 211, 121, 252, 89, 24, 57, 15,
+        255, 175, 12, 201, 179, 23, 247, 117, 169, 254, 227, 206, 220, 210, 169, 217, 42, 172, 102,
+        146, 234, 16, 51, 40, 9, 37, 177, 251, 80, 134, 0, 113, 8, 244, 140, 23, 133, 119, 94, 216,
+        38, 247, 184, 174, 36, 0, 78, 151, 201, 134, 182, 124, 122, 105, 75,
+    ];
+
```
Comment (ArthurHeymans) [inline runtime/tests/integration_tests.rs:951]: I don't think this is the mailbox API is intended. You want to pass a digest of data, not the data itself. The mailbox itself is limited in size and PCRs often get extended with digests of code much larger than the mailbox itself.

File: runtime/tests/integration_tests.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply runtime/tests/integration_tests.rs:927]: I think he just means we should populate `payload_data` with non-zero bytes.

[the remark this answers is not in the document]
Comment (leongross) [reply runtime/tests/integration_tests.rs:927]: @jhand2 my ping was regarding the implementation status of `ureg::MMio`

File: drivers/src/sha384.rs [hunk 14]
```diff
@@ -20,10 +20,10 @@ use crate::{array::Array4x32, wait, Array4x12};
 use caliptra_error::{CaliptraError, CaliptraResult};
 use caliptra_registers::sha512::Sha512Reg;
 
-const SHA384_BLOCK_BYTE_SIZE: usize = 128;
+pub const SHA384_BLOCK_BYTE_SIZE: usize = 128;
```
Comment (jhand2) [inline drivers/src/sha384.rs:23]: Don't think this needs to be public.

File: drivers/src/sha384.rs [hunk 15]
```diff
@@ -20,10 +20,10 @@ use crate::{array::Array4x32, wait, Array4x12};
...
 const SHA384_BLOCK_LEN_OFFSET: usize = 112;
 const SHA384_MAX_DATA_SIZE: usize = 1024 * 1024;
-const SHA384_HASH_SIZE: usize = 48;
+pub const SHA384_HASH_SIZE: usize = 48;
```
Comment (jhand2) [inline drivers/src/sha384.rs:26]: Same. This isn't being used outside this module.

File: drivers/src/lib.rs [hunk 16]
```diff
@@ -42,7 +42,7 @@ mod persistent;
 pub mod printer;
 mod sha1;
 mod sha256;
-mod sha384;
+pub mod sha384;
```
Comment (jhand2) [inline drivers/src/lib.rs:45]: Don't think this is needed

File: drivers/test-fw/src/bin/pcrbank_tests.rs [hunk 17]
```diff
@@ -84,9 +84,12 @@ fn test_write_protection_stickiness() {
     }
 }
 
+fn test_mailbox_pcrextend_cmd() {}
```
Comment (jhand2) [inline drivers/test-fw/src/bin/pcrbank_tests.rs:87]: Assume this will be filled out before merging?

File: common/src/mailbox_api.rs [hunk 18]
```diff
@@ -293,6 +294,28 @@ impl Default for InvokeDpeReq {
...
+    pub const DATA_MAX_SIZE: usize = 48;
+
+    pub fn new(
+        hdr: MailboxReqHeader,
+        pcr_idx: u32,
+        data: [u8; 48],
+    ) -> Result<Self, ExtendPcrReqErr> {
+        Ok(ExtendPcrReq { hdr, pcr_idx, data })
+    }
+}
```
Comment (ArthurHeymans) [inline common/src/mailbox_api.rs:316]: Why have a function for this? It's just filling in a struct with 3 entries with 3 arguments, which is likely inefficient as arguments get copied in functions.

File: runtime/tests/integration_tests.rs [hunk 19]
```diff
@@ -895,3 +899,129 @@ fn test_idev_id_cert() {
...
+    // How to read PCR regs from model?
+    // get pv = PvReg sha512pv.pcr_ctrl().at(id.into()).?
+
+    // prob. similar to drivers/test-fw/src/bin/sha384_tests.rs
+    // fn test_pcr_hash_extend_single_block_3() {
+    //     let mut sha384 = unsafe { Sha384::new(Sha512Reg::new()) };
+    //     let mut pcr_bank = unsafe { PcrBank::new(PvReg::new()) };
+
+    // #[peripheral(offset = 0x1001_8000, mask = 0x0000_7fff)]
+    // pub key_vault: KeyVault,
```
Comment (ArthurHeymans) [inline runtime/tests/integration_tests.rs:992]: Get rid of this commented out code?

File: runtime/tests/integration_tests.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply runtime/tests/integration_tests.rs:927]: I don't think it's necessary to use ureg. Just make `extension_data` non-zero

File: common/src/mailbox_api.rs (hunk 18 above)
Comment (leongross) [reply common/src/mailbox_api.rs:316]: This is an artifact from where the default constructor was not suitable, I will make it compact again.

File: runtime/tests/integration_tests.rs [hunk 20]
```diff
@@ -895,3 +895,96 @@ fn test_idev_id_cert() {
...
+    // for i in 0..32 {
+    //     let e = entry.at(i as usize);
+    //     for ii in 0..12 {
+    //         //         // this fails, but why?
+    //         let ee = e.at(ii).read();
+    //     }
+    // }
+
+    let pcr_4 = pcr.pcr_ctrl().at(4);
+    let r = pcr_4.read();
```
Comment (ArthurHeymans) [inline runtime/tests/integration_tests.rs:956]: Are these used?

**Comment by leongross** [comment]
Can we re-open this PR?

**Comment by jhand2** [comment]
@leongross @ArthurHeymans told me he was taking over the PCR work and would open new PRs. Github isn't letting me reopen this PR ("The mailbox_extend_pcr was force-pushed or recreated"). Could you talk with Arthur and one of you open a new PR with these changes?

Sorry about the confusion.
